README.md
# Sweet Nothings - The Powerful Whisper and Chat Module

This module leverages Foundry v9's new Keybinding API to register a simple, configurable keybinding that brings up a simple yet powerful dialog option to make whispering and chat messages flexible!

## Usage

By default, the keybinding is ALT-W.  Pressing this will present the Whisper Sweet Nothings dialog box.  There are two modes - Whisper, and Chat.  Whisper is the default, and you can check as many recipients as you want.

As a nice shortcut, you may use ALT-R to reply to the last whisper you have received.  This will automatically select the person who sent the whisper.

Alternatively, you can swap to the Chat mode, which also allows you to set the various chat modes allowed by Foundry.  In Character (IC) and Emote chat modes will also generate the chat bubble above your token's head.  If you are the GM, this will also set the speaker to your selected token.  Utilizing the "Other" chat type will result as speaking as your user, rather than any owned or controlled token.

The angled arrows on the left of the dialog allow you to open/close the whisper history, which will show up to the last week's worth of whispers.  This list will automatically filter 
the whisper history based on the selected recipient(s).

## Configuration

Sweet Nothings main configuration is through Foundry's Configure Controls dialog, under the Settings Sidebar.  Sweet Nothings registers it's own category, and you can modify the keybinding shortcut here.

You can also define the defaults for the dialog, either through the normal settings sidebar or using the cog at the top of the Whisper Sweet Nothings dialog.

![The Whisper Dialog](https://gitlab.com/geekswordsman/SweetNothings/-/wikis/uploads/16a37434911ceef2d838d4eaa9350870/Sweet_Nothings_Whisper_Dialog_1_3.webp)

![The Chat Dialog](https://gitlab.com/geekswordsman/SweetNothings/-/wikis/uploads/a9d2f527bf42d62f561cff27c1ddfdfe/Sweet_Nothings_Chat_Dialog.JPG)

![Configuration Dialog](https://gitlab.com/geekswordsman/SweetNothings/-/wikis/uploads/39e0f053cd987f6548876a676e2ee188/Sweet_Nothings_Config_1_3_1.webp)

![Sample Output](https://gitlab.com/geekswordsman/SweetNothings/-/wikis/uploads/f2acfee538cdbd00a6870f55910dcdf6/Sweet_Nothings_Sample.JPG)

## Support Me!
This module is free for use, however if you enjoy it and wish to help support the developer, please visit my Ko-Fi and leave a tip! [Geekswordsman's Ko-Fi!](https://ko-fi.com/geekswordsman)

## Attribution
- 'ting.ogg' notification sound created by Aesterial-Arts originally titled 'Notification Ping', available at https://freesound.org/people/Aesterial-Arts/sounds/633876/ (CC BY 4.0)
- 'triple_ting.ogg' notification sound created by wassapp, originally titled 'wassap_notification', available at https://freesound.org/people/Tundidor/sounds/634505/ (CC BY 3.0)
- 'comm_trill.ogg' notification sound created by TMPZ_1, originally titled 'NotificationLoop', available at https://freesound.org/people/TMPZ_1/sounds/449238/ (CC BY 3.0)
- 'chord_descend.ogg' notification sound created by ebcrosby, title same, available at https://freesound.org/people/ebcrosby/sounds/332960/ (CC BY 3.0)
- 'digital_chime.ogg' notification sound created by Jofae, originally titled 'Chime Notification', available at https://freesound.org/people/Jofae/sounds/380482/ (CC0 1.0 Universal)
- 'computer_chime.ogg' notification sound created by TheAtomicBrain, originally titled 'Computer Chimes - Notification', available at https://freesound.org/people/TheAtomicBrain/sounds/351879/ (CC0 1.0 Universal)
- 'dinging_chime.ogg' notification sound created by sleepyhead16, originally titled 'strange notification', available at https://freesound.org/people/sleepyhead16/sounds/565907/ (CC0 1.0 Universal)
- 'notification_ding' notificatnio sound by Daphne_in_Wonderland, originally titled 'Messenger Notification Sound Imitation', with attirbution to "Notification Pop" by elmasmalo1 of Freesound.org, available at https://freesound.org/people/Daphne_in_Wonderland/sounds/400697/ (CC BY 3.0)

CHANGELOG.md
# Sweet Nothings
## v1.5.0 - 2023-07-24
 - Fixed bug that should have been resolved in v1.4.3 with whispers not being sent when editor set to ProseMirror ( #16 )
 - Fixed bug that caused notifications to be played globally rather than locally ( #19 )
 - Fixed bug that caused no notifications unless "for rolls" was enabled ( #20 )
 
## v1.4.3 - 2023-07-13
 - Fixed bug that prevented whispers from being sent when editor was set to ProseMirror

## v1.4.2 - 2023-05-31
 - Fixed deprecation warnings
 - Fixed bug that could cause the whisper dialog box to not render.
 - Fixed bug where whispering only to a GM could cause the whisper to fail.
 - Fixed bug that would close the dialog if Send button was clicked but no whisper recipients.
 - Fixed bug that did not properly set the engine for the text editor based on configuration choice.
 - Added ability to send whispers as the selected token.
 
## v1.4.1 - 2023-05-24
 - FoundryVTT v11 Compatibility Update
 - Fixed bug where Reply To function was not properly working
 
## v1.4.0 - 2022-09-15
 - Enabled ability to change the Rich Text Editor, making it either Prose (Foundry VTT v10 Default), TinyMCE (Foundry VTT v9 Default), or None (basic text box)
 - Reworked whisper history, vastly improving dialog load and response times! ( #12 )
 - Added a small API for the module, available via game.whisperSweetNothings.  Currently only one method - game.whisperSweetNothings.whisperSweetNothings(isReply)
 
## v1.3.4 - 2022-07-21
 - FoundryVTT v10 Compatibility Update
 
## v1.3.3 - 2022-06-30
 - Created a framework for enabled/disabling audio and toast notifications for external modules
 - Added the "damage-log" as an external module to above framework.
 - Adjusted the border styling of the whisper history to make it more visible when over certain other UI's.

## v1.3.2 - 2022-06-08
 - Fixed a bug where chat modes were not respecting the selected token.
 
## v1.3.1 - 2022-06-08
 - Added new configuration options to enable toast, sound, and appearance in whisper history for messages with roll data (disabled by default)

## v1.3.0 - 2022-06-05
 - Corrected bug where config menu could not be accessed with the cog wheel from the Sweet Nothings dialog.
 - Corrected bug where, upon changing preferred defaults, they would not take until a refresh happened.
 - Corrected bug where, when using the reply function, it was not correctly picking the correct recipient
 - Changed reply logic to allow sending whispers to any user, whether they were online or not.
 - Added an option for GM's to force whispers to always be tagged as In Character, to play nice with the Tabbed Chatlog module.
 - Added a global, per client option to show a toast notification when they received a whisper.
 - Added a global, per client configurable option to play a sound when receiving a new whisper.
 - Added a new one-time message upon a user signing in, introducing them to (Whisper) Sweet Nothings.
 - Added a whisper history that filters based on the selected recipient(s).

## v1.2.2 - 2022-03-14
 - Added Japanese translation thanks to @besardida!
  
## v1.2.1 - 2022-02-21
 - Removed unnecessary console.log
 - Fixed keybind event handlers to prevent default browser behaviors from also firing
  
## v1.2.0 - 2022-01-12
 - Added Reply functionality
 
## v1.0.0 - 2021-12-27
- First release of Sweet Nothings

module/sweetnothings.mjs
import { SweetNothingsAPI } from "./api.mjs";
import { SWEETNOTHINGS } from "./config.mjs";
import { SweetNothingsConfig } from "./forms/sweetNothingsConfig.mjs";
import { SweetNothingsDialog } from "./forms/sweetNothingsDialog.mjs";
import { AboutDialog } from "./about/about-dialog.mjs";
import { Logger } from "./logger/logger.mjs";

export class SweetNothings {
    static initialize() {
        Logger.info(true, `Initializing ${game.i18n.localize("SWEETNOTHINGS.ABOUT.Title")}`);
        AboutDialog.TITLE = game.i18n.localize("SWEETNOTHINGS.ABOUT.Title");

        //Currently cannot localize keybindings in v9.238, so name and hint are temporarily not localized.
        game.keybindings.register(SWEETNOTHINGS.ID, "whisperSweetNothings", {
            name: "Name",
            hint: "Hint",
            editable: [
                {
                    key: "KeyW",
                    modifiers: ["Alt"]
                }
            ],
            onDown: () => {  },
            onUp: () => { this.whisperSweetNothings(); },
            restricted: false,
            precedence: CONST.KEYBINDING_PRECEDENCE.NORMAL
        });

        game.keybindings.register(SWEETNOTHINGS.ID, "whisperSweetNothingsReply", {
            name: "Name",
            hint: "Hint",
            editable: [
                {
                    key: "KeyR",
                    modifiers: ["Alt"]
                }
            ],
            onDown: () => {  },
            onUp: () => { this.whisperSweetNothingsReply(); },
            restricted: false,
            precedence: CONST.KEYBINDING_PRECEDENCE.NORMAL
        })

        //Game Settings
        const sweetSettings = Object.keys(SWEETNOTHINGS.SETTINGS);
        for (let setting of sweetSettings) {
            game.settings.register(SWEETNOTHINGS.ID, setting, SWEETNOTHINGS.SETTINGS[setting]);
        }

        game.settings.register(SWEETNOTHINGS.ID, "GM_ALWAYS_IC", {
            name: game.i18n.localize("SWEETNOTHINGS.CONFIGURATION.GM_ALWAYS_IC.Name"),
            hint: game.i18n.localize("SWEETNOTHINGS.CONFIGURATION.GM_ALWAYS_IC.Hint"),
            restricted: true,
            config: true,
            default: true,
            scope: 'world',
            type: Boolean
        })

        game.settings.register(SWEETNOTHINGS.ID, "WhisperToastNotification", {
            restricted: false,
            default: true,
            scope: 'client',
            type: Boolean
        });

        game.settings.register(SWEETNOTHINGS.ID, "WhisperEnableSound", {
            restricted: false,
            scope: 'client',
            type: Boolean,
            default: true
        });

        game.settings.register(SWEETNOTHINGS.ID, "WhisperNotificationSound", {
            restricted: false,
            scope: 'client',
            type: String,
            default: 'modules/sweetnothings/sounds/ting.ogg'
        });

        game.settings.register(SWEETNOTHINGS.ID, "WhisperNotificationVolume", {
            restricted: false,
            scope: 'client',
            type: Number,
            default: 0.1
        });

        game.settings.register(SWEETNOTHINGS.ID, "WhisperHistoryLength", {
            restricted: false,
            scope: 'client',
            type: Number,
            default: 7
        });

        game.settings.register(SWEETNOTHINGS.ID, "WhisperRollToastNotification", {
            restricted: false,
            scope: 'client',
            type: Boolean,
            default: false
        });

        game.settings.register(SWEETNOTHINGS.ID, "WhisperRollSoundNotification", {
            restricted: false,
            scope: 'client',
            type: Boolean,
            default: false
        });

        game.settings.register(SWEETNOTHINGS.ID, "WhisperRollInHistory", {
            restricted: false,
            scope: 'client',
            type: Boolean,
            default: true
        });

        game.settings.register(SWEETNOTHINGS.ID, "ExternalModuleNotifications", {
            restricted: false,
            scope: 'client',
            type: Object,
            default: {
                "damage-log": {
                    "audio": false,
                    "toast": false
                }
            }
        });

        //Register the setting menu
        game.settings.registerMenu(SWEETNOTHINGS.ID, "UserConfiguration", {
            name: "PerUserConfiguration",
            label: game.i18n.localize("SWEETNOTHINGS.CONFIGURATION.LABEL"),
            hint: game.i18n.localize("SWEETNOTHINGS.CONFIGURATION.HINT"),
            icon: "fas fa-wrench",
            type: SweetNothingsConfig
        });

        /* Detect External Modules */
        for (let key of Object.keys(SWEETNOTHINGS.EXTERNAL_MODULES)) {
            SWEETNOTHINGS.EXTERNAL_MODULES[key] = !!game.modules.get(key)?.active;
        }

        //Handlebar helper for localizing settings
        Handlebars.registerHelper("sweetNothingsLocalizer", function(key) {
            return game.i18n.localize(`SWEETNOTHINGS.CONFIGURATION.${key}`);
        });

        Handlebars.registerHelper('sweetNothingsSelected', function(value1, value2) {
            if (value1 === value2) {
                return 'selected';
            }

            return '';
        });

        Handlebars.registerHelper('sweetNothingsChecked', function(value1, value2) {
            if (value1 === value2) {
                return 'checked';
            }

            return '';
        });

        return this.preloadHandlebarTemplates();
    }

    static ready() {
        SWEETNOTHINGS.FOUNDRY_VERSION = game.version ?? game.data.version;

        let sweetActions = game.keybindings.actions.get("sweetnothings.whisperSweetNothings");
        sweetActions.name = game.i18n.localize("SWEETNOTHINGS.TITLE");
        sweetActions.hint = game.i18n.localize("SWEETNOTHINGS.HINT");

        sweetActions = game.keybindings.actions.get("sweetnothings.whisperSweetNothingsReply");
        sweetActions.name = game.i18n.localize("SWEETNOTHINGS.KEYBINDINGS.REPLY.NAME");
        sweetActions.hint = game.i18n.localize("SWEETNOTHINGS.KEYBINDINGS.REPLY.HINT");

        SweetNothings._checkGreetings();
        $(document).on("click",".sweetNothingsInitialConfig", SweetNothings._configureSweetWhispers.bind(this));

        //Init API
        game.whisperSweetNothings = SweetNothingsAPI;
    }

    static preloadHandlebarTemplates = async function() {
        const templates = Object.keys(SWEETNOTHINGS.TEMPLATES).map(function(key) { return SWEETNOTHINGS.TEMPLATES[key]; });
        return loadTemplates(templates);
    }

    static whisperSweetNothings() {
        let dialog = new SweetNothingsDialog();
        dialog.display();
        return true;
    }

    static whisperSweetNothingsReply() {
        let dialog = new SweetNothingsDialog();
        dialog.display(true);
        return true;
    }

    static checkForWhisper(message) {
        let enableToastNotification = game.settings.get(SWEETNOTHINGS.ID, "WhisperToastNotification");
        let enableToastRollNotification = game.settings.get(SWEETNOTHINGS.ID, "WhisperRollToastNotification");
        let enableNotificationSound = game.settings.get(SWEETNOTHINGS.ID, "WhisperEnableSound");
        let enableNotificationRollSound = game.settings.get(SWEETNOTHINGS.ID, "WhisperRollSoundNotification");

        const configurations = game.settings.get(SWEETNOTHINGS.ID, "ExternalModuleNotifications");

        if ((SWEETNOTHINGS.FOUNDRY_VERSION >= 10 && message.whisper && message.whisper.includes(game.userId)) || (SWEETNOTHINGS.FOUNDRY_VERSION < 10 && message.data.whisper && message.data.whisper.includes(game.userId))) {
            //This message is a whisper to us!
            let hasRollInfo = false;
            if (SWEETNOTHINGS.FOUNDRY_VERSION >= 10) {
                hasRollInfo = Array.isArray(message.rolls) && message.rolls.length > 0;
            } else {
                hasRollInfo = message.data.roll !== undefined;
            }
            let showNotification = hasRollInfo ? (enableToastRollNotification && enableToastNotification) : enableToastNotification;
            let playSound = hasRollInfo ? (enableNotificationRollSound && enableNotificationSound) : enableNotificationSound;

            //Check for module settings
            for (let key of Object.keys(configurations)) {
                if ((SWEETNOTHINGS.FOUNDRY_VERSION >= 10 && message.flags[key]) || (SWEETNOTHINGS.FOUNDRY_VERSION < 10 && message.data.flags[key])) {
                    showNotification = configurations[key].toast;
                    playSound = configurations[key].audio;
                }
            }

            if (message.author === game.userId) {
                if (!hasRollInfo) {
                    showNotification = playSound = false;
                }
            }

            if (showNotification) {
                let sender = ChatMessage.getSpeaker(message).alias;
                ui.notifications.info(game.i18n.localize("SWEETNOTHINGS.NOTIFICATIONS.NEW_MESSAGE") + ` ${sender}`);
            }

            if (playSound) {
                let src = game.settings.get(SWEETNOTHINGS.ID, "WhisperNotificationSound");
                let volume = game.settings.get(SWEETNOTHINGS.ID, "WhisperNotificationVolume");
                AudioHelper.play({src, volume, autoplay: true, loop: false}, false);
            }
        }
    }

    static async _checkGreetings() {
        let initialGreeting = game.user.getFlag(SWEETNOTHINGS.ID, "WhisperGreetingCompleted");
        if (!initialGreeting) {
            let content = await renderTemplate(SWEETNOTHINGS.TEMPLATES.GREETING, {});

            let chatData = {
                author: game.userId,
                content,
                type: CONST.CHAT_MESSAGE_TYPES.WHISPER,
                whisper: [game.userId],
                speaker: null
            };
    
            let message = await ChatMessage.create(chatData, { });

            game.user.setFlag(SWEETNOTHINGS.ID, "WhisperGreetingCompleted", true);
        }
    }

    static _configureSweetWhispers() {
        let config = new SweetNothingsConfig();
        config.render(true);
    }
}

module/workers/chatWorker.mjs
onmessage = function(settings) {
    this.self.parseMessages(settings.data);
};

async function parseMessages(settings) {
    let days = settings.historyLength;
    let today = new Date();
    let filter = new Date(today.getFullYear(), today.getMonth(), today.getDate()-days).getTime();
    let includeRollMessages = settings.includeRolls;
    let targets = settings.targets;

    let baseMessages = null;
    baseMessages = settings.messages.filter(m => m.timestamp >= filter && m.whisper.includes(settings.userId)).sort((a, b) => { return a.timestamp > b.timestamp ? -1 : 1; });
    if (!includeRollMessages) { baseMessages = baseMessages.filter(m => m.rolls === undefined); }

    let toRender = [];
    //Filter now based on selected targets
    if (targets && targets.length > 0 && !targets.includes('GM')) {
        for (let target of targets) {
            toRender = settings.fvttGeneration >= 10 ? toRender.concat(baseMessages.filter(m => m.user === target || m.whisper.includes(target))) : toRender.concat(baseMessages.filter(m => m.data.user === target || m.data.whisper.includes(target)));
        }
    } else {
        toRender = toRender.concat(baseMessages);
    }

    if (toRender.length > 500) {
        //Too many to go through, let's cut it down to size
        toRender = toRender.slice(0,500);
    }

    postMessage(toRender);
}

module/forms/sweetNothingsConfig.mjs
import { SWEETNOTHINGS } from "../config.mjs";
import { HelpFormApplication } from "../about/help-form-application.mjs";

export class SweetNothingsConfig extends HelpFormApplication {
    constructor(object, options) {
        if (!object) { object = {} };
        object.enableAboutButton = true;

        super(object, options);
    }

    static get defaultOptions() {
        const options = super.defaultOptions;

        const overrides = {
            id: "sweetNothingsConfig",
            template: SWEETNOTHINGS.TEMPLATES.CONFIG,
            submitOnChange: true,
            closeOnSubmit: false,
            width: 500,
            height: "auto",
            title: game.i18n.localize("SWEETNOTHINGS.CONFIGURATION.LABEL")
        };

        return foundry.utils.mergeObject(options, overrides);
    }

    activateListeners(html) {
        super.activateListeners(html);

        html.on('click', '[data-action]', this._handleButtonClick.bind(this));
    }

    _onSelectFile(selection, filePicker) {
        this.submit();
    }

    async getData(options) {
        let sweetKeys = Object.keys(SWEETNOTHINGS.SETTINGS);
        let sweetSettings = {};
        let enableSound = game.settings.get(SWEETNOTHINGS.ID, "WhisperEnableSound");
        let enableNotification = game.settings.get(SWEETNOTHINGS.ID, "WhisperToastNotification");
        let notificationSound = game.settings.get(SWEETNOTHINGS.ID, "WhisperNotificationSound");
        let notificationVolume = game.settings.get(SWEETNOTHINGS.ID, "WhisperNotificationVolume");
        let historyLength = game.settings.get(SWEETNOTHINGS.ID, "WhisperHistoryLength");
        let enableRollInHistory = game.settings.get(SWEETNOTHINGS.ID, "WhisperRollInHistory");
        let enableRollToastNotification = game.settings.get(SWEETNOTHINGS.ID, "WhisperRollToastNotification");
        let enableRollSoundNotification = game.settings.get(SWEETNOTHINGS.ID, "WhisperRollSoundNotification");

        /* External Module Configurations */
        const configurations = game.settings.get(SWEETNOTHINGS.ID, "ExternalModuleNotifications");

        let externalModules = [];
        for (let key of Object.keys(SWEETNOTHINGS.EXTERNAL_MODULES)) {
            externalModules.push({
                id: key,
                label: game.modules.get(key)?.data?.title,
                enabled: SWEETNOTHINGS.EXTERNAL_MODULES[key],
                enableToast: configurations[key].toast,
                enableAudio: configurations[key].audio
            });
        }

        for (let key of sweetKeys) {
            let currentValue = game.settings.get(SWEETNOTHINGS.ID, key);

            sweetSettings[key] = {
                value: currentValue,
                choices: SWEETNOTHINGS.SETTINGS[key].choices
            }
        }

        return { sweetSettings, enableSound, enableNotification, notificationSound, notificationVolume, historyLength, enableRollInHistory, enableRollToastNotification, enableRollSoundNotification, externalModules };
    }

    async _updateObject(event, formData) {
        let configurations = game.settings.get(SWEETNOTHINGS.ID, "ExternalModuleNotifications");
        let keys = Object.keys(formData);
        for (let key of keys) {
            if (!key.startsWith("external_")) {
                if (key === "DEFAULT_MESSAGE_ENGINE" && !(game.release?.generation >= 10)) {
                    if (formData[key] === "prosemirror") { formData[key] === "tinymice"; }
                }
                await game.settings.set(SWEETNOTHINGS.ID, key, formData[key]);
            } else {
                let externalKey = "";
                if (key.endsWith("_toast")) {
                    externalKey = key.replace("external_", "").replace("_toast", "");
                    configurations[externalKey].toast = formData[key];
                } else {
                    externalKey = key.replace("external_", "").replace("_audio", "");
                    configurations[externalKey].audio = formData[key];
                }
            }
        }

        await game.settings.set(SWEETNOTHINGS.ID, "ExternalModuleNotifications", configurations);

        this.render();
    }

    async _handleButtonClick(event) {
        const clickedElement = $(event.currentTarget);
        const action = clickedElement.data().action;

        switch (action) {
            case 'playSound':
                let src = game.settings.get(SWEETNOTHINGS.ID, "WhisperNotificationSound");
                let volume = game.settings.get(SWEETNOTHINGS.ID, "WhisperNotificationVolume");
                AudioHelper.play({src, volume, autoplay: true, loop: false}, false);
                break;
            default:
                break;
        }
    }
}

module/config.mjs
export const SWEETNOTHINGS = {};

SWEETNOTHINGS.ID = "sweetnothings";
SWEETNOTHINGS.MODULE_PATH = `modules/${SWEETNOTHINGS.ID}/`;
SWEETNOTHINGS.FOUNDRY_VERSION = 0;

SWEETNOTHINGS.SETTINGS = {
    "DEFAULT_CHATMODE": {
        config: false,
        default: 'other',
        scope: 'client',
        type: String,
        choices: {
            "ic": "SWEETNOTHINGS.DIALOG.MODES.IC",
            "ooc": "SWEETNOTHINGS.DIALOG.MODES.OOC",
            "emote": "SWEETNOTHINGS.DIALOG.MODES.EMOTE",
            "other": "SWEETNOTHINGS.DIALOG.MODES.OTHER"
        }
    },
    "DEFAULT_DIALOG": {
        config: false,
        default: 'whisper',
        scope: 'client',
        type: String,
        choices: {
            "whisper": "SWEETNOTHINGS.DIALOG.WHISPER",
            "chat": "SWEETNOTHINGS.DIALOG.CHAT"
        }
    },
    "DEFAULT_WHISPER_SOURCE": {
        config: false,
        default: 'user',
        scope: 'client',
        type: String,
        choices: {
            "user": "SWEETNOTHINGS.DIALOG.SOURCEUSER",
            "token": "SWEETNOTHINGS.DIALOG.SOURCETOKEN"
        }
    },
    "WHISPER_HISTORY_LENGTH": {
        config: false,
        default: 7,
        scope: 'client',
        type: String,
        choices: {
            1: "1 Day",
            5: "5 Days",
            7: "7 Days",
            14: "14 Days",
            30: "30 Days"
        }
    },
    "DEFAULT_MESSAGE_ENGINE": {
        config: false,
        default: game.release?.generation >= 10 ? "prosemirror" : "tinymce",
        scope: 'client',
        type: String,
        choices: {
            "prosemirror": "Default (Prose Mirror, FVTT v10+)",
            "tinymce": "TinyMCE (Deprecated/FVTT v9)",
            "none": "None"
        }
    }
}

SWEETNOTHINGS.TEMPLATE_PATH = `${SWEETNOTHINGS.MODULE_PATH}/templates/`;
SWEETNOTHINGS.TEMPLATES = {
    CONFIG: `${SWEETNOTHINGS.TEMPLATE_PATH}sweetNothingsConfig.hbs`,
    DIALOG: `${SWEETNOTHINGS.TEMPLATE_PATH}sweetNothingsDialog.hbs`,
    GREETING: `${SWEETNOTHINGS.TEMPLATE_PATH}sweetNothingsGreeting.hbs`,
    HISTORY: `${SWEETNOTHINGS.TEMPLATE_PATH}sweetNothingsHistory.hbs`
}

SWEETNOTHINGS.EXTERNAL_MODULES = {
    "damage-log": false
}